examples/bridge-app/linux/include/Clusters.h

```diff
@@ -0,0 +1,371 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/reporting/reporting.h>
+#include <app/util/attribute-storage.h>
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <functional>
+
+static constexpr uint16_t ZCL_DESCRIPTOR_CLUSTER_REVISION           = 1;
+static constexpr uint16_t ZCL_BRIDGED_DEVICE_BASIC_CLUSTER_REVISION = 1;
+static constexpr uint16_t ZCL_FIXED_LABEL_CLUSTER_REVISION          = 1;
+static constexpr uint16_t ZCL_ON_OFF_CLUSTER_REVISION               = 4;
+static constexpr uint16_t ZCL_SWITCH_CLUSTER_REVISION               = 1;
+static constexpr uint16_t ZCL_LEVEL_CONTROL_CLUSTER_REVISION        = 1;
+
+// This is the interface to cluster implementations, providing access to manipulate attributes.
+class ClusterImpl
+{
+public:
+    virtual ~ClusterImpl() = default;
+
+    virtual void SetEndpointId(chip::EndpointId id)                                                                          = 0;
+    virtual chip::ClusterId GetClusterId()                                                                                   = 0;
+    virtual EmberAfStatus Read(const EmberAfAttributeMetadata * attributeMetadata, uint8_t * buffer, uint16_t maxReadLength) = 0;
+    virtual EmberAfStatus Write(const EmberAfAttributeMetadata * attributeMetadata, uint8_t * buffer)                        = 0;
+    virtual void WriteFromBridge(chip::AttributeId attributeId, const uint8_t * buffer)                                      = 0;
+    virtual chip::Span<const EmberAfAttributeMetadata> GetAllAttributes()                                                    = 0;
+};
+
+// *Type classes are used by code generation to specialize storage for a particular underlying type.
+
+// This provides storage for a primitive binary type, eg uintN_t, float, double
+// Values in ValueType are native, values in byte buffers are LE.
+template <typename T, size_t Bytes, EmberAfAttributeType AfType>
```

Comment:
Why are we re-creating the type space for converting data to/from the Ember storage format, when we already have `cluster-objects.h` to handle this more scalably to/from TLV?

The Ember format is not what we should be using to solve this problem, given its in-ability to deal with structs and lists.

Reply:
Updated to new interface. Definitions in cluster-objects.h are suitable for reading/writing but do not all provide storage